### ISSUE TITLE
check BU RAM disk usage

### ISSUE THREAD

**Issue by andreh12** [cmsdaq/DAQExpert#32]
The BU RAM disk occupancy should be checked against some threshold value for each BU.

We have the corresponding occupancy already in the class `rcms.utilities.daqaggregator.data.BU`.

**Comment by andreh12** [comment]
`evb::bu::ResourceManager::updateResources(..)` in https://github.com/mommsen/evb blocks all 'resources' (FUs) when the (relative) RAM disk occupancy reaches the high watermark (i.e. it effectively will stop processing events).

By default the high watermark is set to 95% and the low watermark to 75%.

**Comment by gladky** [comment]
It's a matter of adding small LM, it's small issue, I mark it to do just before resuming the operation for version 2.3

**Comment by mommsen** [comment]
The logic for blocking or throttling of the BU is much more involved. The whole logic is described here:
https://twiki.cern.ch/twiki/bin/view/CMS/CDAQOnCallHowTOs#BU_blocks_throttles_or_shows_fun

In most circumstances, the BUs block long before the RAM disk occupancy reaches 75%. I think we should come up with some logic based on the BU states for the expert to make shifters aware of potential problems with the FFF.

**Comment by andreh12** [comment]
I agree with @mommsen that we can rely on states of the BUs to alert the shifter. Nevertheless, it would be good to also indicate to the shifter what the origin of the problem could be, i.e. why the BUs went into `Blocked` state. 

As a first step, it could be useful to know (in the flashlist) which of the reasons listed in the twiki page above triggered the `Blocked` state. 

Other than that I can think of the following reasons why BUs can become slow:
* BU died completely 
* too many FUs died
* network problem between BU and FUs
* HLT code crashes for many events
* HLT code takes too much time for many events
* HLT output rate is too high due to bad/wrong HLT menu
* HLT output rate is too high due to too high L1 rate (which in turn could be wrong prescale column applied, detector noise, wrong L1 menu)
* network problem between BU and merger nodes
* merger node died

Many of these underlying reasons could be detected separately (and should therefore be displayed to the shifter with higher priority).

**Comment by mommsen** [comment]
Hi,

I created a new EvB version 5.4.0, where I added a new field 'statusMsg' to the BU flashlist. This field contains a human readable status message, which should explain why the BU is throttling or blocking requests. This information is also visible on the hyperdaq page of the BU.

I think it would be useful if the DAQExpert could display this message whenever a BU is in Throttled/Blocked/Mist/Cloud state. However, it is not clear to me how to display this message when multiple BUs (for maybe different reasons) are in these states.

This new EvB version has been deployed in daq2val, and pending deployment in cDAQ.

**Comment by andreh12** [comment]
As I learned from @gladky recently, the method `register(..)` of class `rcms.utilities.daqexpert.reasoning.base.Context` basically adds a string to be displayed for a given placeholder to a list, i.e. multiple calls to `register(..)` will result in multiple strings being displayed for the placeholder. 

In other words: one could just display all messages of BUs in a bad state, maybe after some prioritization of displaying only `Blocked` BUs if there is at least one in this state, then checking `Throttled` ones, then `Cloud` and finally `Mist` ?

**Comment by mommsen** [comment]
This is a possibility. However, you risk that you display 73 messages in case that all BUs go into a bad state, e.g. if there is a systematic issue on the HLT farm. Maybe one should limit this to 5 or so messages with an indicator that there are more problems.

BTW: I would treat Blocked/Cloud and Throttled/Mist at the same level. The only difference is if there are FUs allocated to the cloud or not.

**Comment by andreh12** [comment]
if the messages were static (i.e. do not contain problem specific or BU specific parts), it would 
be easy to group and count them.

Looking at the corresponding commit mommsen/evb@e7a2659 it looks like this is not the case but in principle one could catch them with regexes (and even sort them according to the value in the status message, e.g. report those with the highest ram disk usage above threshold first). 

But then I see that there can be more than one reason in a message...

**Comment by mommsen** [comment]
Hi @andreh12,

a given BU can indeed throttle for different reasons, which all enter into the load-balancing equation. E.g. a BU which has less FUs but the same RAMdisk usage will throttle more than one which has more FUs with the same RAMdisk usage. If we want to really do some statistics, I guess we should add a bit field where the BU sets the bits according to the cases which are active. However, I don't think that it is really useful for the shifter to see that 2 BUs throttle because they have less FUs, another 5 because they have also a high RAMdisk usage, another 3 because the output b/w is too high and another 6 which in addition have too many LS  queued, etc.

Cheers,
Remi

**Comment by andreh12** [comment]
Hello @mommsen,

I agree that in this case it is too much to show all the details to the shifter but we still have to decide what to display. What message should be displayed in this case ? 

I would vote for the one which has highest occurrence (i.e. 6 which have too many LS queued in your example) so that by understanding / solving this problem one has fixed as many BUs as possible in one go.

**Comment by mommsen** [comment]
Another option would be to display only the longest message. This most likely will contain the cause which bothers other BUs, too.

**Comment by mommsen** [comment]
The 'statusMsg' from the BUs are now also available in cDAQ and miniDAQ.

**Comment by andreh12** [comment]
For the moment we agreed to just show the longest message(s). To have a more machine readable format of the detailed causes we opened an issue here: evb#9

**Comment by andreh12** [comment]
here is an example case of when all FUs crashed due to a problem with the HLT configuration: `snapshots/pro/cdaq/2017/5/27/5/1495861515760.smile`. One sees for example:
```
    "numFUsHLT" : 0,
    "numFUsCrashed" : 34624,
    "numFUsStale" : 0,
    "numFUsCloud" : 0,
```
and all 57 BUs report:

```
Caught exception: exception::FFF 'All FU cores in the appliance are quarantined, i.e. HLT is failing on all of them.'
```

@mommsen can you confirm that in the new scheme from mommsen/evb#9 this would correspond to all BUs reporting `ALL_FUS_QUARANTINED` ?

**Comment by mommsen** [comment]
@andreh12, yes correct.

**Comment by andreh12** [comment]
I see now the field `statusKeywords` in the BU flashlist on daqval which we can now include in class `rcms.utilities.daqaggregator.data.BU` (e.g. as a set of String). 

I noticed that in the same class we do not have a field corresponding to the flashlist column `statusMsg` which contains a human readable description of what the BU is doing. 

Do we want to the human readable description string ? I'm asking because on one hand it would be useful to display to the shifter, on the other hand this column is filled even if everything is fine and this may increase the needed storage by a bit (however the smile compression may make it such that the string only needs to be stored once per snapshot for the first BU if all BUs are in the same state).

**Comment by andreh12** [comment]
Alternatively, we could just store the `statusMsg` when `statusKeywords` is non-empty (to save space). 

This could be confusing when looking at snapshots in the past but if the assumption holds that there is no problem on the BUs when `statusKeywords` is empty, should be ok.

Any opinions ?

**Comment by andreh12** [comment]
http://daq-expert-dev.cms/daq2view-react/index.html?setup=testbed&time=2017-06-06-10:54:38
is a good example/test case where all 57 BUs were in state `ALL_FUS_QUARANTINED` and consequently the `SCAL` FED at some point went into 100% busy.

(the flashlists corresponding to this point in time are `/daqexpertflashlists/flashlists-dev/persistence/testbed/flashlists/*/2017/6/6/8/1496739278035.smile`)
